Add some emoji spinners

nnmrts opened the pull request:
I played around with some emoji and emoji sequences, and this is the result.

sindresorhus commented on spinners.json:
```diff
@@ -1181,5 +1181,165 @@
+			"🖐",
+			"✋",
+			"🤚",
+			"👆"
```
Seems like these are not all the same rendered width. I tried it in `example.js` and the `Unicorn` text moves.

nnmrts replied:
I couldn't try it out in my shell yet, but at least [in the browser it looks fine](https://jsfiddle.net/sg2whe7t/), is this a shell-only thing, that different emoji have different width?

nnmrts replied:
Okay, so for some reason ☝️ and ✌️ are not shown in their "emoji form", in my shell at least. Removing them and adding a space after the 🖐 one fixed it for me.

![Kapture 2020-06-28 at 8 02 44](https://user-images.githubusercontent.com/20396367/85939434-e70c0d80-b915-11ea-96ef-8b8966d13a79.gif)

nnmrts replied:
I've ultimately removed these emoji, so this seems to be resolved.

sindresorhus commented on spinners.json:
```diff
@@ -1181,5 +1181,165 @@
 			"ββββββρ"
 		]
+	},
+	"fingerdance": {
```
```suggestion
	"fingerDance": {
```

sindresorhus commented on spinners.json:
```diff
@@ -1181,5 +1181,165 @@
+			"🤜    🤛",
+			"🤜    🤛",
+			" 🤜  🤛",
+			"  🤜🤛",
```
The width should be constant.

sindresorhus replied:
This one is awesome though!

nnmrts replied:
Hmmm, I'm really struggling with this one. Is there a unicode character for a space with the width of an emoji?

sindresorhus commented on spinners.json:
```diff
@@ -1181,5 +1181,165 @@
+			"🧑    ⚽️    🧑",
+			"🧑     ⚽️   🧑",
+			"🧑      ⚽️  🧑",
+			"🧑       ⚽️🧑",
```
Constant width.

Can you write a test for that? To ensure that all the frames in each spinner has the same width.

nnmrts replied:
Adding a test for constant width could be pretty hard, if not impossible. Look at this small example of the fingerDance spinner:

![Kapture 2020-06-28 at 8 57 29](https://user-images.githubusercontent.com/20396367/85940588-95678100-b91d-11ea-8084-55ed478355ad.gif)

No matter which package or method I use, the number is never the same across all frames, even though visually every frame has the same width.

I used this code to log this btw:

<details>
<summary>code</summary>

```js
const logUpdate = require('log-update');
const cliSpinners = require('.');
const stringWidth = require("string-width");
const stringLength = require("string-length");

const spinner = cliSpinners[process.argv[2] || 'dots'];
let i = 0;

setInterval(() => {
	const {frames} = spinner;
	logUpdate(
		'\n' +
		frames[i = ++i % frames.length] + ' Unicorns' + 
		'\n' +
		'js length:\t\t' + frames[i].length +
		'\n' +
		'"string-width" length:\t' + stringWidth(frames[i]) +
		'\n' + 
		'"string-length" length:\t' + stringLength(frames[i])
	);
}, spinner.interval);
```


</details>

sindresorhus replied:
> Adding a test for constant width could be pretty hard, if not impossible. Look at this small example of the fingerDance spinner:

If the problem is just with that spinner, we could exclude it from that test.

nnmrts replied:
Excluding the spinner from the test would've been somewhat of a solution, although the "fix" I had above doesn't work for every terminal I tested. I've removed the specific emoji causing the issue and added a test, so I think this is resolved as well.

sindresorhus commented on spinners.json:
```diff
@@ -1181,5 +1181,165 @@
+			"🤜 ✨ 🤛"
+		]
+	},
+	"header": {
```
I think the name could be better and more specific.

nnmrts replied:
Alright, I'll change it to "soccerHeader".

sindresorhus commented:
These are all very nice! Thanks for making them. 🙌

nnmrts commented on spinners.json:
```diff
@@ -1181,5 +1181,165 @@
+			"🧑  ⚽️      🧑"
+		]
+	},
+	"mindblown": {
+		"interval": 160,
+		"frames": [
+			"😐",
+			"😐",
+			"😮",
+			"😮",
+			"😦",
+			"😦",
+			"😧",
+			"😧",
+			"🤯",
+			"💥",
+			"✨",
+			"",
+			"",
+			""
+		]
+	},
```
I'm also not sure about that one tbh. Yeah, it's "just" emoji and "just" a cli-spinner, but it looks pretty graphic and insensible. I think the same about the already implemented "grenade" spinner btw.

sindresorhus replied:
I like this one. It's fun.

nnmrts replied:
Okay then :)

sindresorhus commented:
If it was not clear, this is waiting for https://github.com/sindresorhus/cli-spinners/pull/46#discussion_r446803883

sindresorhus commented:
@nnmrts Bump :)

nnmrts commented:
@sindresorhus, ye, sorry, I'll look into it again tonight.

sindresorhus commented:
@nnmrts Bump :)

nnmrts commented:
tl;dr everything should be fixed now, some things were removed because they are dumb, the grenade spinner was updated as well because its first two frames were shorter and i added spaces at the end of all the emoji frames because the existing emoji spinners also have one space at the end; these spinners all have constant widths **in a terminal**, but not always in browsers; if that's an issue i'll remove those ones too

---

So, revisiting this, I don't really know how to continue in a sound way. Excuse my language but it seems like string and character width handling in the context of emoji is a complete clusterfuck. Different browsers, different terminals, different programming languages and different libraries all do it differently. I can't even visually compare these spinners in my IDE because VSCode **also** thinks for some reason it's a good idea that emoji have a slightly larger width than any other characters in an environment that specifically uses fixed-width characters, Terminus and the macOS Terminal thinks that too, Hyper on the other hand doesn't. All this probably changes with different fonts as well. This is literally untestable using normal spaces and emoji.

Anyways, considering this, and that this projects targets are mainly terminals, I added a constant width test using `string-length` that seems to be solid at checking the visual constant width of strings for all terminals. I also replaced "normal" spaces with [ideographic spaces](https://en.wikipedia.org/wiki/Whitespace_character#Unicode) where it's necessary. I copy-pasted them literally, because then you can actually visually see and compare the real lengths of these strings. This may lead to confusion though, I can also change them to `\u+3000` to make it clearer these aren't normal spaces, if you want me to.

Of course this wasn't enough, for some reason ☝️, ✌️ and 🖐️ don't have the usual emoji width in the terminals I tested, they are about a normal space width shorter. They have the usual width in VSCode and browsers though. They also aren't [ZWJ sequences](https://emojipedia.org/emoji-zwj-sequence/) or [keycap sequences](https://emojipedia.org/emoji-keycap-sequence/). I now removed those emoji from that spinner. This makes no sense at all though. At least with emoji like 1️⃣, it makes some sense that they behave different, because they are keycap sequences, but it still doesn't make sense that their emoji representations have completely different widths.

Oh shoot, did I mention that the math spinner is also completely broken and unfixable? I didn't? Well, the math spinner is also completely broken and unfixable. One could theoretically implement a parity test for keycap sequences (if there's an even number of these fuckup emoji in a string, they cancel each other out and the resulting string width should be predictable again), but I just called it a day at this point and completely removed this beautiful math spinner. Sad.

I never want to do this again, I regret this whole PR and I wish all the implementers of all of these text editors and displayers all the luck in the world figuring this out.

Thanks for coming to my ted talk.

sindresorhus commented on test.js:
```diff
@@ -1,8 +1,29 @@
+
+test('constant width', t => {
+	for (const key of Object.keys(cliSpinners)) {
+		if (key !== 'default') {
```
```suggestion
		if (key !== 'default') {
			continue;
		}
```

to reduce nesting.

sindresorhus commented:
Oh wow. Haha. I'm sorry you had to suffer through that...

sindresorhus commented:
> I can also change them to \u+3000 to make it clearer these aren't normal spaces, if you want me to.

👍  Yes

sindresorhus commented:
Can you fix the merge conflict?

nnmrts commented:
> Can you fix the merge conflict?

I hope I did this right. I don't know when the last time was I resolved merge conflicts through GitHub.

But the tests seem to pass, everything seems fine, I think this is it. Yay!

sindresorhus commented:
Woot! 👍🏻

SirWindfield commented:
The math spinner was really nice :( Maybe I'll go through the suffering to get it back into the package.
